src/expr/core/src/expr/expr_udf.rs

```diff
@@ -221,3 +246,49 @@ pub(crate) fn get_or_create_client(link: &str) -> Result<Arc<ArrowFlightUdfClien
         Ok(client)
     }
 }
+
+/// Get or create a wasm runtime.
+///
+/// Runtimes returned by this function are cached inside for at least 60 seconds.
+/// Later calls with the same link will reuse the same runtime.
+#[cfg_or_panic(not(madsim))]
+pub async fn get_or_create_wasm_runtime(link: &str) -> Result<Arc<WasmRuntime>> {
+    // link -> (runtime, last_used_time)
+    type Map = HashMap<String, (Arc<WasmRuntime>, Instant)>;
+    static RUNTIMES: LazyLock<tokio::sync::Mutex<Map>> = LazyLock::new(Default::default);
+
+    let mut runtimes = RUNTIMES.lock().await;
+    let now = Instant::now();
+    // evict unused runtimes for the past 60 seconds
+    runtimes.retain(|l, (rt, last_used_time)| {
+        l == link || Arc::strong_count(rt) > 1 || now.duration_since(*last_used_time).as_secs() < 60
+    });
+    if let Some((runtime, last_used_time)) = runtimes.get_mut(link) {
+        // reuse existing runtime
+        *last_used_time = now;
+        return Ok(runtime.clone());
+    }
```

Comment:
"Moka" appears to be a good fit here.

https://docs.rs/moka/0.12.2/moka/sync/struct.CacheBuilder.html#method.support_invalidation_closures
https://docs.rs/moka/0.12.2/moka/sync/struct.CacheBuilder.html#method.time_to_idle